### ISSUE TITLE
[QIR] QIR generation fails when updating UDT named items in same operation as a conditional return

### ISSUE THREAD

**Issue by cgranade** [microsoft/qsharp-compiler#957]
**Describe the bug**

The following Q# snippet using both named item copy-and-updates and a `return` inside an `if`/`else` block fails to compile when QIR generation is enabled:

**To Reproduce**

```xml
<Project Sdk="Microsoft.Quantum.Sdk/0.15.2103133969">

    <PropertyGroup>
      <OutputType>Exe</OutputType>
      <TargetFramework>netcoreapp3.1</TargetFramework>
      <QirGeneration>true</QirGeneration>
      <QscVerbosity>D</QscVerbosity>
    </PropertyGroup>
  
</Project>
```

```qsharp
namespace ConditionalReturns {
    open Microsoft.Quantum.Arrays;
    open Microsoft.Quantum.Random;
    open Microsoft.Quantum.Intrinsic;
    open Microsoft.Quantum.Canon;

    newtype HasString = (
        Value: String,
        Data: Double[]
    );

    newtype HasDouble = (
        Value: Double
    );

    @EntryPoint()
    operation RunMain() : Unit {
        let x = DrawRandomDouble(0.0, 1.0);
        let c = HasDouble(DrawRandomDouble(0.45, 0.55));
        let s = HasString("s", [x]);
        let a = Zero;
        Message($"{ReturnsConditionally(a, s, c)}");
    }

    operation ReturnsConditionally<'T>(a : 'T, x : HasString, c : HasDouble) : HasString {
        mutable z = [0.0];
        let s = HasString("<", [0.0]);
        // Works if the following is commented out:
        let updated = s w/ Data <- [0.1, 0.2];
        if Head(x::Data) <= c::Value {
            return s;
        } else {
            return s;
        }
        // Works if the following is uncommented:
        // return s;
    }

}
```

Command-line output:

```
PS> dotnet build
Microsoft (R) Build Engine version 16.9.0-preview-21103-02+198f3f262 for .NET
Copyright (C) Microsoft Corporation. All rights reserved.

  Determining projects to restore...
  All projects are up-to-date for restore.
  You are using a preview version of .NET. See: https://aka.ms/dotnet-core-preview
  (1,1): info: Compiling with command line arguments
      ResponseFiles: obj/qsharp/config/qsc.rsp
      OutputFolder: (null)
      ProjectName: (null)
      EmitDll: False
      QirOutputFolder: (null)
      PerfOutputFolder: (null)
      TrimLevel: 1
      Plugins:
      TargetSpecificDecompositions:
      ExposeReferencesViaTestNames: False
      AdditionalAssemblyProperties:
      RuntimeCapabilityName: (null)
      MakeExecutable: False
      Verbosity: D
      OutputFormat: MsBuild
      Input:
      CodeSnippet: (null)
      WithinFunction: False
      References:
      NoWarn: System.Int32[]
      PackageLoadFallbackFolders:
  (1,1): info: Loaded rewrite steps that are executing as part of the compilation process
      (none)
  (1,1): info: Compiling source files
      C:\Users\cgran\source\scratch\qcvv-scratch\bug-repros\conditional-returns\ConditionalReturns.qs
  (1,1): info: Compiling with referenced assemblies
      C:\Users\cgran\.nuget\packages\numsharp\0.20.5\lib\netstandard2.0\NumSharp.Core.dll
      C:\Users\cgran\.nuget\packages\microsoft.quantum.standard\0.15.2103133969\lib\netstandard2.1\Microsoft.Quantum.Standard.dll   
      C:\Users\cgran\.nuget\packages\microsoft.quantum.targets.interfaces\0.15.2103133969\lib\netstandard2.1\Microsoft.Quantum.Targets.Interfaces.dll
      C:\Users\cgran\.nuget\packages\microsoft.quantum.runtime.core\0.15.2103133969\lib\netstandard2.1\Microsoft.Quantum.Runtime.Core.dll
      C:\Users\cgran\.nuget\packages\microsoft.quantum.qsharp.foundation\0.15.2103133969\lib\netstandard2.1\Microsoft.Quantum.QSharp.Foundation.dll
      C:\Users\cgran\.nuget\packages\microsoft.quantum.simulators\0.15.2103133969\lib\netstandard2.1\Microsoft.Quantum.Simulation.QCTraceSimulatorRuntime.dll
      C:\Users\cgran\.nuget\packages\microsoft.quantum.qsharp.core\0.15.2103133969\lib\netstandard2.1\Microsoft.Quantum.QSharp.Core.dll
      C:\Users\cgran\.nuget\packages\microsoft.quantum.simulators\0.15.2103133969\lib\netstandard2.1\Microsoft.Quantum.Simulation.Common.dll
      C:\Users\cgran\.nuget\packages\microsoft.quantum.simulators\0.15.2103133969\lib\netstandard2.1\Microsoft.Quantum.Simulators.dll
  : info: 
  System.InvalidOperationException: cannot close scope that has pending calls to increase reference counts
     at Microsoft.Quantum.QsCompiler.QIR.ScopeManager.CloseScope(Boolean isTerminated) in D:\a\1\s\submodules\qsharp-compiler\src\QsCompiler\QirGeneration\ScopeManager.cs:line 499
     at Microsoft.Quantum.QsCompiler.QIR.GenerationContext.EndFunction() in D:\a\1\s\submodules\qsharp-compiler\src\QsCompiler\QirGeneration\Context.cs:line 561
     at Microsoft.Quantum.QsCompiler.QIR.QirNamespaceTransformation.OnProvidedImplementation(QsTuple`1 argTuple, QsScope body) in D:\a\1\s\submodules\qsharp-compiler\src\QsCompiler\QirGeneration\Subtransformations\NamespaceTransformation.cs:line 76
     at Microsoft.Quantum.QsCompiler.Transformations.Core.NamespaceTransformationBase.OnSpecializationImplementation(SpecializationImplementation implementation) in D:\a\1\s\submodules\qsharp-compiler\src\QsCompiler\Core\Transformations\NamespaceTransformation.fs:line 185
     at Microsoft.Quantum.QsCompiler.Transformations.Core.NamespaceTransformationBase.OnSpecializationKind(QsSpecialization spec) in D:\a\1\s\submodules\qsharp-compiler\src\QsCompiler\Core\Transformations\NamespaceTransformation.fs:line 194
     at Microsoft.Quantum.QsCompiler.QIR.QirNamespaceTransformation.OnSpecializationDeclaration(QsSpecialization spec) in D:\a\1\s\submodules\qsharp-compiler\src\QsCompiler\QirGeneration\Subtransformations\NamespaceTransformation.cs:line 82
     at Microsoft.FSharp.Collections.Internal.IEnumerator.map@75.DoMoveNext(b& curr) in E:\A\_work\130\s\src\fsharp\FSharp.Core\seq.fs:line 78
     at Microsoft.FSharp.Collections.Internal.IEnumerator.MapEnumerator`1.System-Collections-IEnumerator-MoveNext() in E:\A\_work\130\s\src\fsharp\FSharp.Core\seq.fs:line 64
     at System.Collections.Generic.LargeArrayBuilder`1.AddRange(IEnumerable`1 items)
     at System.Collections.Generic.EnumerableHelpers.ToArray[T](IEnumerable`1 source)
     at System.Linq.Enumerable.ToArray[TSource](IEnumerable`1 source)
     at System.Collections.Immutable.ImmutableArray.CreateRange[T](IEnumerable`1 items)
     at Microsoft.Quantum.QsCompiler.Transformations.Core.NamespaceTransformationBase.OnCallableKind(QsCallable c) in D:\a\1\s\submodules\qsharp-compiler\src\QsCompiler\Core\Transformations\NamespaceTransformation.fs:line 240
     at Microsoft.Quantum.QsCompiler.QIR.QirNamespaceTransformation.OnCallableDeclaration(QsCallable c) in D:\a\1\s\submodules\qsharp-compiler\src\QsCompiler\QirGeneration\Subtransformations\NamespaceTransformation.cs:line 96
     at Microsoft.Quantum.QsCompiler.Transformations.Core.NamespaceTransformationBase.OnNamespaceElement(QsNamespaceElement element) in D:\a\1\s\submodules\qsharp-compiler\src\QsCompiler\Core\Transformations\NamespaceTransformation.fs:line 294
     at Microsoft.FSharp.Collections.Internal.IEnumerator.map@75.DoMoveNext(b& curr) in E:\A\_work\130\s\src\fsharp\FSharp.Core\seq.fs:line 78
     at Microsoft.FSharp.Collections.Internal.IEnumerator.MapEnumerator`1.System-Collections-IEnumerator-MoveNext() in E:\A\_work\130\s\src\fsharp\FSharp.Core\seq.fs:line 64
     at Microsoft.FSharp.Collections.SeqModule.Iterate[T](FSharpFunc`2 action, IEnumerable`1 source) in E:\A\_work\130\s\src\fsharp\FSharp.Core\seq.fs:line 494
     at Microsoft.Quantum.QsCompiler.Transformations.Core.NamespaceTransformationBase.OnNamespace(QsNamespace ns) in D:\a\1\s\submodules\qsharp-compiler\src\QsCompiler\Core\Transformations\NamespaceTransformation.fs:line 317
     at Microsoft.Quantum.QsCompiler.QIR.Generator.Apply() in D:\a\1\s\submodules\qsharp-compiler\src\QsCompiler\QirGeneration\Generator.cs:line 61
     at Microsoft.Quantum.QsCompiler.BuiltInRewriteSteps.QirGeneration.Transformation(QsCompilation compilation, QsCompilation& transformed) in D:\a\1\s\submodules\qsharp-compiler\src\QsCompiler\Compiler\RewriteSteps\QirGeneration.cs:line 75
     at Microsoft.Quantum.QsCompiler.LoadedStep.Transformation(QsCompilation compilation, QsCompilation& transformed) in D:\a\1\s\submodules\qsharp-compiler\src\QsCompiler\Compiler\LoadedStep.cs:line 209
     at Microsoft.Quantum.QsCompiler.CompilationLoader.ExecuteRewriteStep(LoadedStep rewriteStep, QsCompilation compilation, QsCompilation& transformed) in D:\a\1\s\submodules\qsharp-compiler\src\QsCompiler\Compiler\CompilationLoader.cs:line 913

EXEC(1,1): error QS7109: The compilation step "QIR Generation" loaded from "C:\Users\cgran\.nuget\packages\microsoft.quantum.sdk\0.15.2103133969\tools\qsc\Microsoft.Quantum.QsCompiler.dll" threw an exception. [C:\Users\cgran\source\scratch\qcvv-scratch\bug-repros\conditional-returns\ConditionalReturns.csproj]
  
  ____________________________________________

  Q# compilation failed: 1 error, 0 warnings
  1 logged exception

C:\Users\cgran\.nuget\packages\microsoft.quantum.sdk\0.15.2103133969\Sdk\Sdk.targets(143,5): error MSB3073: The command "dotnet "C:\Users\cgran\.nuget\packages\microsoft.quantum.sdk\0.15.2103133969\DefaultItems\../tools/qsc/qsc.dll" build --format MsBuild -v D --response-files obj/qsharp/config/qsc.rsp" exited with code -13. [C:\Users\cgran\source\scratch\qcvv-scratch\bug-repros\conditional-returns\ConditionalReturns.csproj]
C:\Users\cgran\.nuget\packages\microsoft.quantum.sdk\0.15.2103133969\Sdk\Sdk.targets(148,5): error : Q# compiler failed with exit code: -13. [C:\Users\cgran\source\scratch\qcvv-scratch\bug-repros\conditional-returns\ConditionalReturns.csproj]

Build FAILED.

EXEC(1,1): error QS7109: The compilation step "QIR Generation" loaded from "C:\Users\cgran\.nuget\packages\microsoft.quantum.sdk\0.15.2103133969\tools\qsc\Microsoft.Quantum.QsCompiler.dll" threw an exception. [C:\Users\cgran\source\scratch\qcvv-scratch\bug-repros\conditional-returns\ConditionalReturns.csproj]
C:\Users\cgran\.nuget\packages\microsoft.quantum.sdk\0.15.2103133969\Sdk\Sdk.targets(143,5): error MSB3073: The command "dotnet "C:\Users\cgran\.nuget\packages\microsoft.quantum.sdk\0.15.2103133969\DefaultItems\../tools/qsc/qsc.dll" build --format MsBuild -v D --response-files obj/qsharp/config/qsc.rsp" exited with code -13. [C:\Users\cgran\source\scratch\qcvv-scratch\bug-repros\conditional-returns\ConditionalReturns.csproj]
C:\Users\cgran\.nuget\packages\microsoft.quantum.sdk\0.15.2103133969\Sdk\Sdk.targets(148,5): error : Q# compiler failed with exit code: -13. [C:\Users\cgran\source\scratch\qcvv-scratch\bug-repros\conditional-returns\ConditionalReturns.csproj]
    0 Warning(s)
    3 Error(s)

Time Elapsed 00:00:05.12
```

**System information**

```
PS> dotnet --list-sdks
3.1.201 [C:\Program Files\dotnet\sdk]
3.1.300 [C:\Program Files\dotnet\sdk]
5.0.100 [C:\Program Files\dotnet\sdk]
5.0.200 [C:\Program Files\dotnet\sdk]
6.0.100-preview.1.21103.13 [C:\Program Files\dotnet\sdk]
```

**Comment by bettinaheim** [comment]
Regarding the 
```
        // Works if the following is uncommented:
        //return s;
```
in the code: it really doesn't work - adding that line merely tricks the compile and the code fails to execute. Without that line, one of the sanity checks in the compiler is (correctly) failing, because for some reason we fail to process a pending ref count increase for `%updated`.

**Comment by bettinaheim** [comment]
Assigning to myself since I know what the issue is, though it is perfectly possible I may introduce a bug when I fix it - it is in one of the trickier parts, since the issue is in the mechanism that ensures that ref count changes are applied "only" (at least that's the goal) when needed.

**Comment by bettinaheim** [comment]
Reopening due to https://github.com/microsoft/qsharp-compiler/pull/989.

**Comment by bettinaheim** [comment]
Fixed with #990.
